Mathlib/Data/Option/Basic.lean

```diff
@@ -3,28 +3,366 @@ Copyright (c) 2017 Mario Carneiro. All rights reserved.
 -/
-import Mathlib.Init.Function
-import Mathlib.Data.Option.Defs
-import Mathlib.Logic.Basic
+import Mathlib.Logic.IsEmpty
+import Mathlib.Logic.Relator
+import Mathlib.Mathport.Rename
+
+/-!
+# Option of a type
+
+This file develops the basic theory of option types.
+
+If `α` is a type, then `Option α` can be understood as the type with one more element than `α`.
+`Option α` has terms `some a`, where `a : α`, and `none`, which is the added element.
+This is useful in multiple ways:
+* It is the prototype of addition of terms to a type. See for example `with_bot α` which uses
+  `none` as an element smaller than all others.
+* It can be used to define failsafe partial functions, which return `some the_result_we_expect`
+  if we can find `the_result_we_expect`, and `none` if there is no meaningful result. This forces
+  any subsequent use of the partial function to explicitly deal with the exceptions that make it
+  return `none`.
+* `Option` is a monad. We love monads.
+
+`Part` is an alternative to `Option` that can be seen as the type of `True`/`False` values
+along with a term `a : α` if the value is `True`.
+
+-/
 
 namespace Option
 
-theorem some_injective (α : Type _) : Function.Injective (@some α) :=
-  fun _ _ => some_inj.mp
+variable {α β γ δ : Type _}
+
+theorem coe_def : (fun a => ↑a : α → Option α) = some :=
+  rfl
+
+#align option.get_or_else Option.getD
+
+@[simp]
+theorem getD_coe (x y : α) : Option.getD (↑x) y = x :=
+  rfl
+
+theorem coe_get {o : Option α} (h : o.isSome) : ((Option.get _ h : α) : Option α) = o :=
+  Option.some_get h
+
+theorem eq_of_mem_of_mem {a : α} {o1 o2 : Option α} (h1 : a ∈ o1) (h2 : a ∈ o2) : o1 = o2 :=
+  h1.trans h2.symm
+
+theorem Mem.left_unique : Relator.LeftUnique ((· ∈ ·) : α → Option α → Prop) :=
+fun _ _ _=> mem_unique
+
+theorem some_injective (α : Type _) : Function.Injective (@some α) := fun _ _ => some_inj.mp
 
 /-- `option.map f` is injective if `f` is injective. -/
 theorem map_injective {f : α → β} (Hf : Function.Injective f) : Function.Injective (Option.map f)
-| none, none, _ => rfl
-| some a₁, some a₂, H => by rw [Hf (Option.some.inj H)]
+  | none, none, _ => rfl
+  | some a₁, some a₂, H => by rw [Hf (Option.some.inj H)]
+
+@[simp]
+theorem map_comp_some (f : α → β) : Option.map f ∘ some = some ∘ f :=
+  rfl
+
+@[simp]
+theorem none_bind' (f : α → Option β) : none.bind f = none :=
+  rfl
+
+@[simp]
+theorem some_bind' (a : α) (f : α → Option β) : (some a).bind f = f a :=
+  rfl
+
+theorem bind_eq_some' {x : Option α} {f : α → Option β} {b : β} :
+    x.bind f = some b ↔ ∃ a, x = some a ∧ f a = some b :=
+  by cases x <;> simp
+
+theorem bind_eq_none' {o : Option α} {f : α → Option β} :
+    o.bind f = none ↔ ∀ b a, a ∈ o → b ∉ f a := by
+  simp only [eq_none_iff_forall_not_mem, mem_def, bind_eq_some, not_exists, not_and, iff_self]
+
+-- FIXME: there is no global `mjoin`
+-- theorem join_eq_join : mjoin = @join α :=
+--   funext fun x => by rw [mjoin, bind_id_eq_join]
```

Comment:
I guess we need to port `Init/Control/Combinators.lean` as well.

Reply:
#530